include verbose "reject-details" field in testmempoolaccept response

pinheadmz opened the pull request:
Adds a new field `reject-details` in `testmempoolaccept` responses to include `m_debug_message` from `ValidationState`. This string is the complete error message thrown by the mempool in response to `sendrawtransaction`.

The extra verbosity is helpful to consumers of `testmempoolaccept`, which is sort of a debug tool anyway. 

example:
> 
> {
>   "txid": "07d7a59a7bdad4c3a5070659ea04147c9b755ad9e173c52b6a38e017abf0f5b8",
>   "wtxid": "5dc243b1b92ee2f5a43134eb3e23449be03d1abb3d7f3c03c836ed0f13c50185",
>   "allowed": false,
>   "reject-reason": "insufficient fee",
>   "reject-details": "insufficient fee, rejecting replacement 07d7a59a7bdad4c3a5070659ea04147c9b755ad9e173c52b6a38e017abf0f5b8; new feerate 0.00300000 BTC/kvB <= old feerate 0.00300000 BTC/kvB"
> }

pinheadmz committed:
rpc: include verbose reject-details field in testmempoolaccept response

pinheadmz committed:
test: cover testmempoolaccept debug-message in RBF test

pinheadmz committed:
doc: add release note about testmempoolaccept debug-message

luke-jr commented:
Why not pass the details in `m_debug_message` and add that as a new field?

pinheadmz commented:
> Why not pass the details in `m_debug_message` and add that as a new field?

My thinking was to match exactly the single combined string returned by `sendrawtransaction`. But if adding the new field gets your ACK @luke-jr I'll make the change ;-)

glozow reviewed (commented):
Concept ACK to providing the debug string, I'm sure it'd be helpful for users. Would be nice to do this for submitpackage as well.

I'm not sure if we try to keep strings for `reject-reason` stable at all. If we do, I'd be in favor of adding a new field `error_msg` with the full string instead of a separate field with just the debug message.

pinheadmz commented:
> I'm not sure if we try to keep strings for `reject-reason` stable at all.

You mean like how block reject reasons are defined [in BIP22?](https://github.com/bitcoin/bips/blob/master/bip-0022.mediawiki#user-content-Appendix_Example_Rejection_Reasons)

jonatack reviewed (commented):
lgtm ACK bcdc3646c3e24930433e30b6363ca2d6f6fb9198 modulo the question of a new field vs the appending to the `reject_reason` output.

pinheadmz commented:
> Yeah exactly. Like if somebody has a fee bumper that's triggered by `reject-reason == "mempool min fee not met"` or something.

Yeah ok so, the detailed message is appended to the original so "somebody" would just need to change their `==` to a "starts with" or "includes" or regex checker. In my opinion, this API-breaking change is why we have major version numbers and release notes. So I'm not sure how we decide whats precious or not. Either way is fine with me.

pinheadmz commented:
> Needs rebase if still relevant

Done 🙏

instagibbs commented:
concept ACK

unfortunately I know people do use reject-reason strings for various informational purposes. Would be a good project to gather those uses and maybe think of better error codes to capture these?

pinheadmz commented:
@instagibbs thanks for taking a look! Not sure what you mean about gathering uses though. Are you agreeing with @luke-jr that the debug messages should have their own field in TMA response object and leave `reject-reason` alone?

instagibbs commented:
> Not sure what you mean about gathering uses though.

messing with the message will cause production uses to have to update their code.

pinheadmz commented:
> Looking at why error codes aren't enough for production uses

I'm working on an [RBF-batcher with CardCoins](https://blog.cardcoins.co/rbf-batching-at-cardcoins-diving-into-the-mempool-s-dark-reorg-forest) and we were getting `insufficient fee` errors in test from `testmempoolaccept()`, which could be reported from two different places. Inspiration for this PR came from that ambiguity.

https://github.com/bitcoin/bitcoin/blob/6e721c923c87abdb8d99674093d08d8c17bc52c2/src/validation.cpp#L983

https://github.com/bitcoin/bitcoin/blob/6e721c923c87abdb8d99674093d08d8c17bc52c2/src/validation.cpp#L1004

kristapsk commented:
Concept ACK

0xB10C commented:
I have a slight preference to keep the `reject-reason` string as is, but like the idea of having more information on rejections. In https://github.com/0xB10C/find-non-standard-tx and for https://bitcoin-data.github.io/non-standard-transactions/ I use the short `reject-reason`.

pinheadmz committed:
rbf: remove unecessary newline at end of error string

pinheadmz commented:
Thanks for taking a look, @0xB10C. I've refactored the PR to add a new field instead of changing anything already in place. Edited PR title and description. I made the tests super clear that the complete string is returned by `submitrawtransaction` whereas its split into two fields in `testmempoolaccept`

0xB10C commented:
I've played around a bit with this in https://github.com/0xB10C/find-non-standard-tx/pull/2. 

I've seen output for the following reasons:
- `too-long-mempool-chain`: e.g.
  - `exceeds descendant size limit for tx 9f95e53da39ece93ceba3cc78a443a1aef0fc656bf921bbd2aee7f6533f0604f [limit: 101000]`
  -  `exceeds ancestor size limit [limit: 101000]`
  -  `2256302dedb181f83bc51248215e51f680f77d07e11ad7e9e34c2e483ad10e7b [limit: 26]`
- `min relay fee not met`: e.g. `0 < 200`

Other reasons like e.g. `scriptpubkey`, `tx-size`, `dust`, `missing-inputs`, `multi-op-return` have an empty debug-message.

I haven't yet taken a closer look at the code.

pinheadmz commented:
> Other reasons like e.g. `scriptpubkey`, `tx-size`, `dust`, `missing-inputs`, `multi-op-return` have an empty debug-message.

debug-message was empty like `""`? I'd expect the key to not be present at all:

```cpp
if (!state.GetDebugMessage().empty()) result_inner.pushKV("debug-message", state.GetDebugMessage());
```

jonatack reviewed (commented):
Concept ACK d2917e7e0b342509d50325cf7f00da15d81b3c65

pinheadmz commented:
Thanks @DrahtBot good bot, have a biscuit.

Rebased on master to fix silent conflict that failed function test. Also updated PR title and description since the name of the new field has changed

glozow commented on src/rpc/mempool.cpp:
```diff
@@ -242,6 +243,7 @@ static RPCHelpMan testmempoolaccept()
                         result_inner.pushKV("reject-reason", "missing-inputs");
                     } else {
                         result_inner.pushKV("reject-reason", state.GetRejectReason());
+                        if (!state.GetDebugMessage().empty()) result_inner.pushKV("reject-details", state.GetDebugMessage());
```
A simpler way to do this might just be `ToString()`.

It ensures the field exists as long as reject-reason exists. That'd be a simpler interface to understand, as the current one depends on very specific internal stuff - I can't really tell you off the top of my head where we have/don't add a debug string.

It does mean the (very short) first part of the error string is repeated in both places, but perhaps that is convenient for users that want the more verbose one.

Also nit, I heard underscores are better for json parsers
```suggestion
                        result_inner.pushKV("reject_details", state.ToString());
```

pinheadmz replied:
Sure I'll use `ToString()` here. Going to leave the hyphens though because there are several other hyphenated fields in the response object already... but I do see lots of underscores in other RPC responses 😬

pinheadmz replied:
So actually, after making this change I discovered several tests fail everywhere `reject-reason` is checked. There will be a lot of diffs like this... let me know if this changes your mind at all about that `if` condition there.

```diff
--- a/test/functional/mempool_accept.py
+++ b/test/functional/mempool_accept.py
@@ -109,7 +109,7 @@ class MempoolAcceptanceTest(BitcoinTestFramework):
         ))
         # ... 0.99 passes
         self.check_mempool_result(
-            result_expected=[{'txid': txid_in_block, 'allowed': False, 'reject-reason': 'txn-already-known'}],
+            result_expected=[{'txid': txid_in_block, 'allowed': False, 'reject-reason': 'txn-already-known', 'reject-details': 'txn-already-known'}],
             rawtxs=[raw_tx_in_block],
             maxfeerate=0.99,
         )
```

instagibbs replied:
@pinheadmz could pop that field just like the check_mempool_result pops fee field?

glozow commented on test/functional/feature_rbf.py:
```diff
@@ -280,22 +295,30 @@ def test_spends_of_conflicting_outputs(self):
+        assert_raises_rpc_error(-26, f"{reject_reason}, {reject_details}", self.nodes[0].sendrawtransaction, tx2_hex, 0)
+
+        res = self.nodes[0].testmempoolaccept(rawtxs=[tx2_hex])[0]
+        assert res["reject-reason"] == reject_reason
```
any reason not to use `assert_equal`?

pinheadmz replied:
fixed thanks

rkrux replied:
Forgot to push?

glozow commented on doc/release-notes-28121.md:
```diff
@@ -0,0 +1,5 @@
+RPC
```
Can be ignored imo, it'd be deleted once copied to release notes draft anyway

glozow reviewed (commented):
Apologies about the late review! Sorry to bikeshed the interface a bit more, the rest looks good.

rkrux reviewed (commented):
I want to test this out, do you mind rebasing it so that the `CMake` build changes are in this as well?

```
➜  bitcoin git:(tma-debug) ✗ bitcoinmakeall
CMake Error: The source directory "/Users/rkrux/projects/bitcoin" does not appear to contain CMakeLists.txt.
Specify --help for usage, or press the help button on the CMake GUI.
make: *** [cmake_check_build_system] Error 1
```

pinheadmz commented:
Rebased on master (+cmake), and then rebased on master again after https://github.com/bitcoin/bitcoin/pull/31112 merged. This PR now adds some extra coverage to the logging in 31112 as well.

rkrux commented on test/functional/feature_rbf.py:
```diff
@@ -319,14 +342,22 @@ def test_new_unconfirmed_inputs(self):
+        reject_details = f"{reject_reason}, replacement {tx2.hash} adds unconfirmed input, idx 1"
+        assert_raises_rpc_error(-26, f"{reject_details}", self.nodes[0].sendrawtransaction, tx2_hex, 0)
+
+
```
Nit: Extra line

rkrux reviewed (commented):
Successful make and functional tests. Partial review, will complete the review soon.

I tested it out on local with 2 cases: passing in an unsigned transaction, and then passing in a tx with incorrect input.

<details>
<summary>Simulated `testmempoolaccept` failures</summary>

```
➜  ~ bitcoinclireg testmempoolaccept "[\"02000000011d7064f723f688af381d4f2d7f1d8c7ea9d1e0199531b636f3e7499d68dd21900000000000ffffffff0178de052a010000001600148d6b22f6e4c0c5dc96529ca5ebc6ad0cde42966800000000\"]"
[
  {
    "txid": "7d7d6f5e1e5fdc0955ac16309636597c5dc9e117950a3aee2203fb60177131fc",
    "wtxid": "7d7d6f5e1e5fdc0955ac16309636597c5dc9e117950a3aee2203fb60177131fc",
    "allowed": false,
    "reject-reason": "mandatory-script-verify-flag-failed (Witness program hash mismatch)",
    "reject-details": "mandatory-script-verify-flag-failed (Witness program hash mismatch), input 0 of 7d7d6f5e1e5fdc0955ac16309636597c5dc9e117950a3aee2203fb60177131fc (wtxid 7d7d6f5e1e5fdc0955ac16309636597c5dc9e117950a3aee2203fb60177131fc), spending 9021dd689d49e7f336b6319519e0d1a97e8c1d7f2d4f1d38af88f623f764701d:0"
  }
]
➜  ~ bitcoinclireg testmempoolaccept "[\"02000000012d7064f723f688af381d4f2d7f1d8c7ea9d1e0199531b636f3e7499d68dd21900000000000ffffffff0178de052a010000001600148d6b22f6e4c0c5dc96529ca5ebc6ad0cde42966800000000\"]"
[
  {
    "txid": "3b9d266452396611cc6843b2a39adc555538a6f7c59e4cbb653f5826756be0a1",
    "wtxid": "3b9d266452396611cc6843b2a39adc555538a6f7c59e4cbb653f5826756be0a1",
    "allowed": false,
    "reject-reason": "missing-inputs"
  }
]
```

</details>

rkrux commented:
> Rebased on master (+cmake), and then rebased on master again after #31112 merged. This PR now adds some extra test coverage to the logging in 31112 as well.

Thank you for rebasing on master, I can test it out on local now.

rkrux commented on test/functional/feature_rbf.py:
```diff
@@ -154,7 +162,14 @@ def test_doublespend_chain(self):
-        assert_raises_rpc_error(-26, "insufficient fee", self.nodes[0].sendrawtransaction, dbl_tx_hex, 0)
+        reject_reason = "insufficient fee"
+        reject_details = f"{reject_reason}, rejecting replacement {dbl_tx.hash}, less fees than conflicting txs; 3.00 < 4.00"
+        assert_raises_rpc_error(-26, f"{reject_details}", self.nodes[0].sendrawtransaction, dbl_tx_hex, 0)
+
+        res = self.nodes[0].testmempoolaccept(rawtxs=[dbl_tx_hex])[0]
```
In 7fc278cedbc5c7cb047887883e1cb3d76f4f6b4d 'include verbose "reject-details" field in testmempoolaccept response'

Ordering nit: Usually as a user, I would end up calling `testmempoolaccept` RPC before calling the `sendrawtransaction` RPC.

rkrux reviewed (approved):
ACK ed33337d830111807afc9feba43854a49dec8838 on top of the [previous review](https://github.com/bitcoin/bitcoin/pull/28121#pullrequestreview-2484720480).

pinheadmz commented on src/rpc/mempool.cpp:
```diff
@@ -245,6 +246,7 @@ static RPCHelpMan testmempoolaccept()
                         result_inner.pushKV("reject-reason", "missing-inputs");
```
Actually in this case `m_reject_reason` would be `"bad-txns-inputs-missingorspent"` and `m_reject_details` would be an empty string, but the RPC response has this hard coded instead (covered by `mempool_accept.py`)

pinheadmz reviewed (commented):
addressed comments by @rkrux

rkrux reviewed (approved):
re-ACK b6f0593f43304f4ff31e8b68558ceeb1b588403c

Reviewed range diff
```
git range-diff ed33337d830111807afc9feba43854a49dec8838...b6f0593f43304f4ff31e8b68558ceeb1b588403c
```

glozow reviewed (commented):
ACK b6f0593f43304f4ff31e8b68558ceeb1b588403c
